### PR TITLE
adxl345: Driver, initial commit for review.

### PR DESCRIPTION
**Pull request by benoit-pereira-da-silva**
1. Please prefix the issue title with the primary package affected. For example,
if this PR fixes periph.io/x/devices/v3/cap1xxx, prefix the PR title with "cap1xxx:".

2. Mention the issue number it fixes or add the details of the changes if it
doesn't have a specific issue. Examples:

- Added basic support for the adxl345 accelerometer.

3. Once integrated, send a PR to https://github.com/periph/cmd to leverage the
new functionality (if relevant).

### CODE REVIEW

File: adxl345/adxl345.go [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+	"encoding/binary"
+	"fmt"
+	"periph.io/x/conn/v3/physic"
+	"periph.io/x/conn/v3/spi"
+)
+
+type Sensitivity byte
+
+const (
+	DeviceID = 0x00 // Device ID, expected to be 0xE5 when using ADXL345
```
Comment (maruel) [inline adxl345/adxl345.go:13]: Please do not export these unless they are meant to be used by the user.

File: adxl345/adxl345.go [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+	// FIFO control
+	FifoCtl    = 0x38 // FIFO control
+	FifoStatus = 0x39 // FIFO status
+
+)
+
+// DefaultSpiFrequency is the default SPI frequency used to communicate with the device.
+var (
+	SpiFrequency = physic.KiloHertz * 50
```
Comment (maruel) [inline adxl345/adxl345.go:62]: 50kHz is extremely slow for SPI?

Please inline these values, no need to have variables.

File: adxl345/adxl345.go [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+	// Convert the spi.Port into a spi.Conn so it can be used for communication.
+	c, err := p.Connect(SpiFrequency, SpiMode, SpiBits)
+	if err != nil {
+		return nil, err
+	}
+	d := &Dev{
+		name: "ADXL345",
+		s:    c,
+	}
+	if o.TurnOnOnStart {
```
Comment (maruel) [inline adxl345/adxl345.go:106]: I don't think it's necessary, it's okay to have users call it manually, or always do it by default.

File: adxl345/adxl345.go [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+// The ADXL345 combines both registers to deliver 16-bit output for each acceleration axis.
+// A similar approach is used for the Y and Z axes. This technique provides higher precision in the measurements.
+func (d *Dev) ReadAndCombine(reg1, reg2 byte) int16 {
+	low, _ := d.Read(reg1)
+	high, _ := d.Read(reg2)
+	return int16(uint16(high)<<8) | int16(low)
+}
+
+// Read reads a 16-bit value from the specified register address.
+func (d *Dev) Read(regAddress byte) (int16, error) {
```
Comment (maruel) [inline adxl345/adxl345.go:175]: I'd recommend to use https://pkg.go.dev/periph.io/x/conn/v3@v3.7.0/mmr#Dev16.ReadUint16 instead and convert back to a Int16.

File: adxl345/example/example.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+	// Use spireg SPI port registry to find the first available SPI bus.
+	p, err := spireg.Open("")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	defer p.Close()
+
+	o := adxl345.DefaultOpts
+	d, err := adxl345.New(p, &o)
```
Comment (maruel) [inline adxl345/example/example.go:30]: d, err := adxl345.New(p, &adxl345.DefaultOpts)

**Review by maruel — COMMENTED** [review]
Thanks for the PR!

File: adxl345/adxl345.go (hunk 1 above)
Comment (benoit-pereira-da-silva) [reply adxl345/adxl345.go:16]: It is a Register it is normal to keep it public for anyone that wants to read it.

File: adxl345/adxl345.go (hunk 2 above)
Comment (benoit-pereira-da-silva) [reply adxl345/adxl345.go]: You are absolutely right. I ve fixed 2Gz , ADXL2345 supports up to 5gz.

File: adxl345/adxl345.go (hunk 3 above)
Comment (benoit-pereira-da-silva) [reply adxl345/adxl345.go]: Removed.

File: adxl345/example/example.go (hunk 5 above)
Comment (benoit-pereira-da-silva) [reply adxl345/example/example.go]: Done

File: adxl345/adxl345.go (hunk 4 above)
Comment (benoit-pereira-da-silva) [reply adxl345/adxl345.go:181]: @maruel Sorry, I m quite new to "periph" and i do not understand

```go
func (d *Dev) Read(regAddress byte) (int16, error) {
	tx := []byte{regAddress | 0x80, 0x00}
	rx := make([]byte, len(tx))
	err := d.s.Tx(tx, rx)
	if err != nil {
		return 0, err
	}
	return int16(binary.LittleEndian.Uint16(rx)), nil
}``` 
this conversion is equivalent to `return int16(uint16(rx[0]) | uint16(rx[1])<<8)`  i feel that using Dev16.ReadUint16 won't be more efficient.

Comment (maruel) [reply adxl345/adxl345.go:181]: It's okay. The point is to reduce the amount of code as much as possible but it's not a big deal.

**Comment by maruel** [comment]
Please fix the CI error.

**Comment by benoit-pereira-da-silva** [comment]
> Please fix the CI error
I m sorry but i can't figure out what is that error 

 ```Initializing: exit status 128:
fatal: cannot mkdir /home/cuong/go/src/github.com/benoit-pereira-da-silva/devices-0: No space left on device
```

**Comment by maruel** [comment]
Oh, I meant https://github.com/periph/devices/actions/runs/7348677329/job/20007433704?pr=66

File: adxl345/adxl345.go [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+
+type Opts struct {
+	ExpectedDeviceID byte        // Expected device ID used to verify that the device is an ADXL345.
+	Sensitivity      Sensitivity // Sensitivity of the device (2G, 4G, 8G, 16G)
+}
+
+// Dev is a driver for the ADXL345 accelerometer
+// It uses the SPI interface to communicate with the device.
+type Dev struct {
+	c    mmr.Dev16
```
Comment (maruel) [inline adxl345/adxl345.go:93]: Remove it if you don't plant to use it.

File: adxl345/adxl345.go [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+func (d *Dev) String() string {
+	return fmt.Sprintf("ADXL345{Sensitivity:%s}", d.sensitivity)
+}
+
+// New creates a new ADXL345 Dev or returns an error.
+// The bus and chip parameters define the SPI bus and chip select to use.
+// The SPI s is configured.
+// The device is turned on.
+// The device is verified to be an ADXL345.
+func New(p spi.Port, o *Opts) (*Dev, error) {
```
Comment (maruel) [inline adxl345/adxl345.go:110]: I'd recommend to name it NewSPI right away so if I2C is added later, it's a non-breaking change.

**Review by maruel — APPROVED** [review]
Two minor things then it's good to go.

File: adxl345/adxl345.go (hunk 6 above)
Comment (benoit-pereira-da-silva) [reply adxl345/adxl345.go]: Done ...  Will try to be a better citizen next time I will use mmr when providing I2C support. This version as been well tested on real hardwares via SPI only, i don't want to provide I2C without verifying it.

File: adxl345/adxl345.go (hunk 7 above)
Comment (benoit-pereira-da-silva) [reply adxl345/adxl345.go]: Done

**Pull request merged**